sql/core_clients_last_seen_raw_v1.sql

```diff
@@ -0,0 +1,62 @@
+  -- Equivalent to, but more efficient than, calling udf_bitmask_range(1, 28)
+CREATE TEMP FUNCTION
+  udf_bitmask_lowest_28() AS (0x0FFFFFFF);
+  --
+CREATE TEMP FUNCTION
+  udf_shift_one_day(x INT64) AS (IFNULL((x << 1) & udf_bitmask_lowest_28(),
+      0));
+  --
+CREATE TEMP FUNCTION
+  udf_combine_days(prev INT64,
+    curr INT64) AS (udf_shift_one_day(prev) + IFNULL(curr,
+      0));
+  --
+WITH
+  _current AS (
+  SELECT
+    -- In this raw table, we capture the history of activity over the past
+    -- 28 days for each usage criterion as a single 64-bit integer. The
+    -- rightmost bit represents whether the user was active in the current day.
+    CAST(TRUE AS INT64) AS days_seen_bits,
+    -- We only trust profile_date if it is within one week of the ping submission,
+    -- so we ignore any value more than seven days old.
+    IF(DATE_DIFF(submission_date, profile_date, DAY) BETWEEN 0 AND 6,
+      1 << DATE_DIFF(submission_date, profile_date, DAY),
+      0) AS days_created_profile_bits,
+    * EXCEPT (submission_date)
+  FROM
+    core_clients_daily_v1
+  WHERE
+    submission_date = @submission_date ),
+  --
+  _previous AS (
+  SELECT
+    * EXCEPT (submission_date)
+  FROM
+    core_clients_last_seen_raw_v1 AS cls
+  WHERE
+    submission_date = DATE_SUB(@submission_date, INTERVAL 1 DAY)
+    -- Filter out rows from yesterday that have now fallen outside the 28-day window.
+    AND udf_shift_one_day(days_seen_bits) > 0)
+  --
+SELECT
+  @submission_date AS submission_date,
+IF
+  (_current.client_id IS NOT NULL,
+    _current,
+    _previous).* REPLACE ( --
+    udf_combine_days(_previous.days_seen_bits,
+      _current.days_seen_bits) AS days_seen_bits,
+    -- We want to base new profile creation date on the first profile_creation_date
+    -- value we observe, so we propagate an existing non-null value in preference
+    -- to a non-null value on today's observation.
+  IF
+    (udf_shift_one_day(_previous.days_created_profile_bits) > 0,
+      udf_shift_one_day(_previous.days_created_profile_bits),
+      _current.days_created_profile_bits) AS days_created_profile_bits)
```

Comment:
> we propagate an existing non-null value in preference to a non-null value on today's observation.

This makes it sound like we use `COALESCE`, so I think this should use `COALESCE`

```diff
-  IF
-    (udf_shift_one_day(_previous.days_created_profile_bits) > 0,
-      udf_shift_one_day(_previous.days_created_profile_bits),
-      _current.days_created_profile_bits) AS days_created_profile_bits)
+  COALESCE(NULLIF(udf_shift_one_day(_previous.days_created_profile_bits),
+      0),
+    _current.days_created_profile_bits) AS days_created_profile_bits)
```

Additional benefit: because sql doesn't guarantee lazy evaluation of conditional statements, we avoid an extra `udf_shift_one_day` call

Reply:
I think we want `days_created_profile_bits` to consistently end up as 0 rather than null.

We could use COALESCE more simply if we removed the IFNULL in udf_shift_one_day (so that function would return null rather than 0) and added back in an explicit IFNULL in combine_days.

This would then look like:

```
  IFNULL(COALESCE(udf_shift_one_day(_previous.days_created_profile_bits),
    _current.days_created_profile_bits), 0) AS days_created_profile_bits
```

Reply:
We can just add one more value to the `COALESCE`, and that avoids the case where `udf_shift_one_day` returns `0` if `_previous.days_created_profile_bits` is `0`
```diff
-  IF
-    (udf_shift_one_day(_previous.days_created_profile_bits) > 0,
-      udf_shift_one_day(_previous.days_created_profile_bits),
-      _current.days_created_profile_bits) AS days_created_profile_bits)
+  COALESCE(NULLIF(udf_shift_one_day(_previous.days_created_profile_bits),
+      0),
+    _current.days_created_profile_bits,
+    0) AS days_created_profile_bits)
```

Reply:
You're my hero.

Reply:
looks like the last 0 is still missing
```suggestion
      _current.days_created_profile_bits,
      0) AS days_created_profile_bits)
```